hal/inc/core_hal.h

```diff
@@ -284,6 +285,100 @@ typedef void(*HAL_Event_Callback)(int event, int flags, void* data);
 
 void HAL_Set_Event_Callback(HAL_Event_Callback callback, void* reserved);
 
+
+#define HAL_CORE_SLEEP_VERSION 2
+
+typedef enum hal_core_wakeup_source_type_t {
+    HAL_CORE_WAKEUP_SOURCE_TYPE_UNKNOWN = 0,
+    HAL_CORE_WAKEUP_SOURCE_TYPE_GPIO = 1,
+    HAL_CORE_WAKEUP_SOURCE_TYPE_ADC = 2,
+    HAL_CORE_WAKEUP_SOURCE_TYPE_DAC = 3,
+    HAL_CORE_WAKEUP_SOURCE_TYPE_RTC = 4,
+    HAL_CORE_WAKEUP_SOURCE_TYPE_LPCOMP = 5,
+    HAL_CORE_WAKEUP_SOURCE_TYPE_UART = 6,
+    HAL_CORE_WAKEUP_SOURCE_TYPE_I2C = 7,
+    HAL_CORE_WAKEUP_SOURCE_TYPE_SPI = 8,
+    HAL_CORE_WAKEUP_SOURCE_TYPE_TIMER = 9,
+    HAL_CORE_WAKEUP_SOURCE_TYPE_CAN = 10,
+    HAL_CORE_WAKEUP_SOURCE_TYPE_USB = 11,
+    HAL_CORE_WAKEUP_SOURCE_TYPE_BLE = 12,
+    HAL_CORE_WAKEUP_SOURCE_TYPE_NFC = 13,
+    HAL_CORE_WAKEUP_SOURCE_TYPE_NETWORK = 14,
+    HAL_CORE_WAKEUP_SOURCE_TYPE_MAX = 0x7FFFF
+} hal_core_wakeup_source_type_t;
+static_assert(sizeof(hal_core_wakeup_source_type_t) == 2, "length of hal_core_wakeup_source_type_t should be 2-bytes aligned.");
+
+/**
+ * Stop mode:
```

Comment:
This mode aims to make user be aware of what peripherals keep running and consuming power in sleep mode and system will fully manage the reset of peripherals to save more power as possible.